kafka-rest/src/main/java/io/confluent/kafkarest/v2/KafkaConsumerManager.java

```diff
@@ -274,33 +303,107 @@ public String createConsumer(String group, ConsumerInstanceConfig instanceConfig
-              // consumer does. When the new consumer is available we may be able to update this
-              // to provide better feedback to the user.
-              Exception responseException = e;
-              if (!(e instanceof RestException)) {
-                responseException = Errors.kafkaErrorException(e);
+    final KafkaConsumerReadTask task = new KafkaConsumerReadTask<KafkaKeyT, KafkaValueT,
+          ClientKeyT, ClientValueT>(
+          state,
+          timeout,
+          maxBytes,
+          new ConsumerWorkerReadCallback<ClientKeyT, ClientValueT>() {
+            @Override
+            public void onCompletion(
+                    List<? extends ConsumerRecord<ClientKeyT, ClientValueT>> records,
+                    Exception e) {
+              KafkaConsumerManager.this.updateExpiration(state);
+              if (e != null) {
+                // Ensure caught exceptions are converted to RestExceptions so the user gets
+                // a nice error message. Currently we don't define any more specific
+                // errors because the old consumer interface doesn't classify the errors well
+                // like the new consumer does. When the new consumer is available we may be
+                // able to update this to provide better feedback to the user.
+                Exception responseException = e;
+                if (!(e instanceof RestException)) {
+                  responseException = Errors.kafkaErrorException(e);
+                }
+                callback.onCompletion(null, responseException);
+              } else {
+                callback.onCompletion(records, null);
               }
-              callback.onCompletion(null, responseException);
-            } else {
-              callback.onCompletion(records, null);
             }
           }
-        }
     );
+    return executor.submit(new RunnableReadTask(new ReadTaskState(task, state, callback)));
+  }
+
+  class RunnableReadTask implements Runnable, Delayed {
+    private final ReadTaskState taskState;
+    private final KafkaRestConfig consumerConfig;
+    private final long started;
+    private final long requestExpiration;
+    // Expiration if this task is waiting, considering both the expiration of the whole task and
+    // a single backoff, if one is in progress
+    private long waitExpiration;
+
+    public RunnableReadTask(ReadTaskState taskState) {
+      this.taskState = taskState;
+      this.started = config.getTime().milliseconds();
+      this.consumerConfig = taskState.consumerState.getConfig();
+      this.requestExpiration = this.started
+              + consumerConfig.getInt(KafkaRestConfig.CONSUMER_REQUEST_TIMEOUT_MS_CONFIG);
+      this.waitExpiration = 0;
+    }
+
+    @Override
+    public void run() {
+      try {
+        log.trace("Executing consumer read task ({})", taskState.task);
+        long backoffTime = config.getTime().milliseconds()
```

Comment:
Hmm.. we compute this prior to the call to `doPartialRead` which can block. The value will be stale when the read returns. I think it would be better to move this into the task not done case following the partial read.

Reply:
ouch. good catch